File: src/main/webapp/tutorials/code/typescript/generics-functions.ts
// Generic Functions

// Generic function with type inference
function wrap<T>(value: T): { data: T } {
    return { data: value };
}

let wrapped = wrap(123);  // Type inferred as { data: number }
console.log(wrapped);

// Multiple type parameters
function pair<T, U>(first: T, second: U): [T, U] {
    return [first, second];
}

let result = pair("Alice", 25);  // [string, number]
console.log(result);

// Generic array operations
function map<T, U>(arr: T[], fn: (item: T) => U): U[] {
    return arr.map(fn);
}

let numbers = [1, 2, 3];
let strings = map(numbers, n => n.toString());
console.log(strings);

File: src/main/webapp/tutorials/code/go/generics-functions.go
package main

import "fmt"

// Generic function that works with any slice
func Map[T, U any](slice []T, fn func(T) U) []U {
	result := make([]U, len(slice))
	for i, v := range slice {
		result[i] = fn(v)
	}
	return result
}

// Generic function with multiple type parameters
func Zip[T, U any](a []T, b []U) []struct {
	First  T
	Second U
} {
	length := len(a)
	if len(b) < length {
		length = len(b)
	}

	result := make([]struct {
		First  T
		Second U
	}, length)
	for i := 0; i < length; i++ {
		result[i] = struct {
			First  T
			Second U
		}{a[i], b[i]}
	}
	return result
}

func main() {
	// Map: Transform []int to []string
	numbers := []int{1, 2, 3, 4, 5}
	strings := Map(numbers, func(n int) string {
		return fmt.Sprintf("Number: %d", n)
	})
	fmt.Println("Mapped:", strings)

	// Map: Transform []string to []int (lengths)
	words := []string{"Go", "Generics", "Rock"}
	lengths := Map(words, func(s string) int {
		return len(s)
	})
	fmt.Println("Lengths:", lengths)

	// Zip: Combine two slices
	names := []string{"Alice", "Bob", "Charlie"}
	ages := []int{25, 30, 35}
	combined := Zip(names, ages)

	fmt.Println("\nZipped:")
	for _, pair := range combined {
		fmt.Printf("%s is %d years old\n", pair.First, pair.Second)
	}
}
